osd: Improved pg-upmap computing speed

Hu-Yuxuan opened the pull request:
This commit primarily optimizes the logic of the try_drop_remap_overfull function in OSDMap, specifically by considering not only the migrating OSD, but also comparing the deviation between the migrating OSD and the target OSD during the process of dropping existing remapping pairs.

Fixes: https://tracker.ceph.com/issues/63137





<!--
  - Please give your pull request a title like

      [component]: [short description]

  - Please use this format for each git commit message:

      [component]: [short description]

      [A longer multiline description]

      Fixes: [ticket URL on tracker.ceph.com, create one if necessary]
      Signed-off-by: [Your Name] <[your email]>

    For examples, use "git log".
-->

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

Hu-Yuxuan committed:
osd/OSDMap: skip upmap-items drop when um_from would become equally overfull

Improves pg-upmap balancer convergence speed on clusters where many OSDs
are mutually overfull and linked by pg_upmap_items -- a common pattern
after a scale-up that increases replica count.

When try_drop_remap_overfull finds an incoming [um_from -> osd] pair on an
overfull osd, it called test_change to check whether dropping the pair would
improve distribution.  If um_from carries the same excess load as osd,
though, dropping just returns the PG to an equally crowded OSD -- test_change
rejects the move anyway.  With many such mutual pairs, the wasted calls add
up to minutes before the balancer gives up.

This adds an early check: skip test_change when um_from would end up at
least as overfull as osd after receiving the PG back.  This eliminates
O(n_cohort * R) wasted test_change calls per balancer pass and reduces
calc_pg_upmaps from minutes to milliseconds in the affected scenario.

Fixes: https://tracker.ceph.com/issues/63137
Signed-off-by: Hu-Yuxuan <1024252105@qq.com>
Co-authored-by: Kefu Chai <k.chai@proxmox.com>

Hu-Yuxuan commented:
jenkins test make check

Hu-Yuxuan commented:
running make check

Hu-Yuxuan commented:
jenkins test make check

Hu-Yuxuan commented:
@batrick Could you please help me review this code optimization for the pg-upmap? I would greatly appreciate your assistance. Thank you very much.

Hu-Yuxuan commented:
> > @batrick Could you please help me review this code optimization for the pg-upmap? I would greatly appreciate your assistance. Thank you very much.
> 
> I'm not the best choice for RADOS. I ping the relevant reviewers. Thanks!

Thank you so much !

Hu-Yuxuan commented:
@ljflores Could you please help me review this code optimization for the pg-upmap? I would greatly appreciate your assistance. Thank you very much.

ljflores commented:
Hi @Hu-Yuxuan, apologies, I will take a look at this.

ljflores commented on src/osd/OSDMap.cc:
```diff
@@ -5799,7 +5800,7 @@ bool OSDMap::try_drop_remap_overfull(
       auto& um_from = um_pair.first;
       auto& um_to = um_pair.second;
-      if (um_to == osd) {
+      if (um_to == osd && osd_deviation[um_from] + 1 < osd_deviation[osd]) {
```
Better readability:
```suggestion
      if ((um_to == osd) && (osd_deviation[um_from] + 1 < osd_deviation[osd])) {
```

Hu-Yuxuan replied:
> Better readability:

I have modified the code according to your suggestions and tested it. Please kindly help me review it again and if there is no problem, please help me merge it.

ljflores replied:
This part looks good

tchaikov replied:
@ljflores hi Laura, i just noticed this change was made based on your advice. but the dominant convention in the existing tree goes the other way, and existing examples in OSDMap.cc follow the style without parenthesis:

https://github.com/ceph/ceph/blob/e133a14de31f9a5e185b6dae80491438ffb716f8/src/osd/OSDMap.cc#L223
https://github.com/ceph/ceph/blob/e133a14de31f9a5e185b6dae80491438ffb716f8/src/osd/OSDMap.cc#L456
https://github.com/ceph/ceph/blob/e133a14de31f9a5e185b6dae80491438ffb716f8/src/osd/OSDMap.cc#L838
https://github.com/ceph/ceph/blob/e133a14de31f9a5e185b6dae80491438ffb716f8/src/osd/OSDMap.cc#L1441
https://github.com/ceph/ceph/blob/e133a14de31f9a5e185b6dae80491438ffb716f8/src/osd/OSDMap.cc#L2221

Hu-Yuxuan commented:
jenkins test make check arm64

Hu-Yuxuan commented:
jenkins test make check arm64

Hu-Yuxuan commented:
jenkins test make check arm64

ljflores reviewed (commented):
@Hu-Yuxuan can you provide some "before and after" results of the optimization?

frittentheke commented:
@Hu-Yuxuan ?

JoshSalomon commented:
Basically looks good, but since this code is so complex (the change is simple, but I am not sure I can easily assess the impact of this change) I would appreciate another eye looking at this.

JoshSalomon commented:
> @Hu-Yuxuan can you provide some "before and after" results of the optimization?

Can you also the functional impact of this change

frittentheke commented:
@JoshSalomon  If @Hu-Yuxuan does not seem to be very active on Github, but he also opened the tracker ticket with details about his findings and reasoning / intentions of the change. 

I'd hate to see such a potential improvement go to waste. Maybe he can be reached via email - https://tracker.ceph.com/users/14519?

Hu-Yuxuan commented:
> @JoshSalomon If @Hu-Yuxuan does not seem to be very active on Github, but he also opened the tracker ticket with details about his findings and reasoning / intentions of the change.
> 
> I'd hate to see such a potential improvement go to waste. Maybe he can be reached via email - https://tracker.ceph.com/users/14519?

In our production environment, using the unoptimized calculation logic can result in very slow computations in certain clusters with extremely complex topologies. For example, when our cluster is deployed across 3 availability zones, the upmap calculation time can reach days. Currently, our online clusters have widely adopted this optimized code, and when using the optimized logic in the aforementioned extreme cases, we can achieve time optimizations of hundreds or even thousands of times, reducing computation time from days to minutes or seconds. However, we have also found that this may lead to the loss of a very small portion of feasible solutions in extreme cases. Since there are actually many feasible solutions for upmap, this does not seem to affect correctness. We will provide a complete test report later. Thank you all very much.

Hu-Yuxuan commented:
> > @JoshSalomon If @Hu-Yuxuan does not seem to be very active on Github, but he also opened the tracker ticket with details about his findings and reasoning / intentions of the change.
> > I'd hate to see such a potential improvement go to waste. Maybe he can be reached via email - https://tracker.ceph.com/users/14519?
> 
> In our production environment, using the unoptimized calculation logic can result in very slow computations in certain clusters with extremely complex topologies. For example, when our cluster is deployed across 3 availability zones, the upmap calculation time can reach days. Currently, our online clusters have widely adopted this optimized code, and when using the optimized logic in the aforementioned extreme cases, we can achieve time optimizations of hundreds or even thousands of times, reducing computation time from days to minutes or seconds. However, we have also found that this may lead to the loss of a very small portion of feasible solutions in extreme cases. Since there are actually many feasible solutions for upmap, this does not seem to affect correctness. We will provide a complete test report later. Thank you all very much.

@JoshSalomon

frittentheke commented:
@JoshSalomon would love to see this reviewed.

frittentheke commented:
@JoshSalomon would you kindly check on this PR. Improvements of the proposed (!) scale should not go to waste if they hold in review.

JoshSalomon commented on src/osd/OSDMap.h:
```diff
@@ -1559,7 +1559,8 @@ class OSDMap {
     std::set<pg_t>& to_unmap,
-    std::map<pg_t, mempool::osdmap::vector<std::pair<int32_t,int32_t>>>& to_upmap
+    std::map<pg_t, mempool::osdmap::vector<std::pair<int32_t,int32_t>>>& to_upmap,
+    std::map<int,float>& osd_deviation
```
I would make this parameter const

tchaikov replied:
agreed. for better maintainability.

frittentheke commented:
@JoshSalomon thanks for the review. Whats next for this PR in terms of testing and merge?

ceph-jenkins commented:
Can one of the admins verify this patch?

frittentheke commented:
@JoshSalomon ?

frittentheke commented:
@ljflores could you take a look at this PR?

frittentheke commented:
@ljflores ?

tchaikov commented on src/osd/OSDMap.cc:
despite that it's convenient to use `operator[]` to access the map, but better off to use `at()`, so our posterity does not need to reason about "what happens if `um` $\notin$ `osd_deviation`?" even this does not happen at all. but `osd_deviation.at()` makes the invariant that `um_from` always $\in$ `osd_deviation` explicit.

tchaikov commented on src/osd/OSDMap.cc:
this deserves a comment explaining that `+1`  is to evaluate the move adding *1* pg back -- to revoke this upmap mapping, it it does not help, it's pointless to do this. i'd suggest adding a comment like:
```c++
  // +1: dropping this pair moves one PG from um_to back to um_from. Skip the
  // drop if it would leave um_from at least as overfull as osd is today. 
```

tchaikov commented on src/osd/OSDMap.cc:
nit, could drop the parentheses around `um_to == osd` and `osd_deviation[um_from] + 1 < osd_deviation[osd]`.

tchaikov reviewed (approved):
lgtm in general, just five minor issues:

1. could use more constness at the function boundary
2. could have better in-code comment explaining `+1`
3. coding style: to drop the parentheses in `if ((um_to == osd) ...`
4. would be better to have a unit test using `TEST_F(OSDMapTest, ...)` modeling after the `BUG_*` tests. but this is far-from a blocker.

tchaikov commented:
@Hu-Yuxuan Yuxuan, thanks for your contribution. sorry it takes so long!

  Thanks again for sending this — the heuristic is sound and addresses a real pain point on skewed clusters. The PR has been open for a while and I'd like to help move it along.

Are you still actively working on it? If so, happy to leave it in your hands and just review.

If you don't have bandwidth right now, would you mind if I picked it up? I'd like to:

- rebase onto current main,
- address the open review notes (Josh's `const map<int,float>&` suggestion, plus an explanatory comment for the +1 step and a tighter commit message),
- add a unit test under OSDMapTest covering the case the new predicate filters out.

I'd of course preserve your `Signed-off-by` and authorship on the original commit, and add `Co-Authored-by` for the follow-up work. Just let me know which way works for you.

What do you think?

tchaikov committed:
test/osd: add perf test for calc_pg_upmaps with mutual overfull upmap pairs

Reproduces the scenario from https://tracker.ceph.com/issues/63137: 500
cohort OSDs all at deviation +2, linked by pg_upmap_items rings with R=5
pairs each.  Before the fix, every incoming pair triggered a test_change
call that got rejected -- calc_pg_upmaps took ~11 minutes.  After, it
exits in ~200ms.

Signed-off-by: Kefu Chai <k.chai@proxmox.com>

tchaikov commented:
changelog:

- added a perf test which reproduces the issue in https://tracker.ceph.com/issues/63137
- added a unit test to verify that we don't put pgs to the OSDs which are already overfull
- pass `osd_deviation` with `const &` when possible.
- add comment to explain what `+1` is for.

tchaikov commented:
@ljflores @JoshSalomon @rzarzynski hi Laura, Josh and Radek, could you take another look?

tchaikov commented:
ok to test

ljflores commented:
RADOS approved: https://tracker.ceph.com/projects/rados/wiki/MAIN#httpstrackercephcomissues76440
